### DISCUSSION TITLE
Calculating distances

### DISCUSSION THREAD

**Discussion by DrRDStanley** [biolab/orange3#5969] [Q&A] [answered]
**What's wrong?**

Hi! I am brand new to GitHub as well as Orange3. And, I am learning a whole lot of once; it almost feels like cognitive overload, but I will do my best.

I am following along in your document **_Introduction to Data Mining (GS-GS-6203) September 2019_** and looking at Lesson 5: Introduction to Hierarchical Clustering in that document; specifically, my query pertains to the calculation of distances. I was working with grades2.tab and wanted to understand a bit more about the various distance metrics. I'm not entirely sure what it means to "normalize" a metric, but let me focus on the cosine metric, a metric that I have learned about in my readings about similarity measures. I know that if two vectors are "identical," then one is a scalar multiple of the others. Two vectors are said to be "similar" if the angle between them is "close to zero." But if one takes the cosine of an angle near zero, the value will be close to 1. Similarly, two vectors that are nearly orthogonal (an angle close to 90 degrees) would have a cosine close to 0.

I used two cases (Bill and Cynthia), changing their respective grades to 0/1 and 1/0.

<img width="325" alt="Screen Shot 2022-05-11 at 2 52 42 PM" src="https://user-images.githubusercontent.com/105386550/167925143-b8299353-0fce-4fa2-96e1-a1064e858393.png">

Clearly, these two vectors are orthogonal; hence, the angle between them is 90 degrees, where the cosine (90) = 0. But Orange produces a distance matrix that looks like:

<img width="171" alt="Screen Shot 2022-05-11 at 2 59 04 PM" src="https://user-images.githubusercontent.com/105386550/167925630-d2ea0bc6-125d-4617-8640-d83ee4854a19.png">

Why does it produce 1 instead of 0? 

Similarly, when I set the "grades" to be identical, ...as in here....

<img width="320" alt="Screen Shot 2022-05-11 at 3 03 29 PM" src="https://user-images.githubusercontent.com/105386550/167926258-1da3c920-2d7a-405e-99f6-bddbc20c76d6.png">

...one should expect that the cos(0) = 1, but I get zero in the distance matrix.

<img width="156" alt="Screen Shot 2022-05-11 at 3 05 06 PM" src="https://user-images.githubusercontent.com/105386550/167926523-47bfe33a-d7f6-4550-b6dd-75716b69e465.png">

I'll end there instead of posing multiple questions, although I would love to have some clearer insight into the normalization of the metrics you have in the dropdown menu in the distances widget.

**How can we reproduce the problem?** 

<img width="587" alt="Screen Shot 2022-05-11 at 3 10 44 PM" src="https://user-images.githubusercontent.com/105386550/167927553-977053ba-e6e2-4655-aeea-62a669ed67e5.png">

<img width="348" alt="Screen Shot 2022-05-11 at 3 11 12 PM" src="https://user-images.githubusercontent.com/105386550/167927566-fe2bc974-ceaa-4fcb-8b69-6c5190d36c0f.png">

<img width="156" alt="Screen Shot 2022-05-11 at 3 05 06 PM" src="https://user-images.githubusercontent.com/105386550/167927531-d2a87584-d909-461f-be26-2d5b720d8450.png">

Last, if I use the Euclidean distance metric for my first case, I should get a measure of sqrt(2), which I do...

<img width="200" alt="Screen Shot 2022-05-11 at 3 16 57 PM" src="https://user-images.githubusercontent.com/105386550/167928401-3d5f4fe9-bec1-401e-82e5-a5c9abf4fa23.png">

This makes sense, but the normalization process puzzles me. Where does the following come from...

<img width="171" alt="Screen Shot 2022-05-11 at 3 18 56 PM" src="https://user-images.githubusercontent.com/105386550/167928782-2036e598-9832-41a2-9fcb-b21db09b00cb.png">

**What's your environment?**

- Operating system: Mac OS 12.3.1
- Orange version: Orange3.32
- How you installed Orange: from orangedatamining.com

With thanks,
RDS

**Comment by ajdapretnar** [comment, marked as answer]
You are referring to cosine similarity, which is indeed the cosine of the angle between vectors. But this is cosine distance, which is 1-cosine(angle). Hence the difference.

For normalization, please refer to https://en.wikipedia.org/wiki/Feature_scaling and also our own documentation, which explains which type of normalization was used: https://orangedatamining.com/widget-catalog/unsupervised/distances/

**Reply by DrRDStanley** [reply]
Thank you so much, Ajda! There is so much to learn and figure out, but I am making progress!
